File: sources/net.sf.j2s.core/test/dev/j2s/org/ultrastudio/LissajousSimple.js
Clazz.declarePackage ("org.ultrastudio");
Clazz.load (["javax.swing.JApplet", "$.JPanel", "javax.swing.event.ChangeListener", "java.awt.GridLayout", "javax.swing.JSlider"], "org.ultrastudio.LissajousSimple", ["java.awt.BasicStroke", "$.BorderLayout", "$.Color", "$.Dimension", "java.awt.geom.Line2D", "javax.swing.BorderFactory"], function () {
c$ = Clazz.decorateAsClass (function () {
this.acceleration = 100;
if (!Clazz.isClassDefined ("org.ultrastudio.LissajousSimple.LisChart")) {
org.ultrastudio.LissajousSimple.$LissajousSimple$LisChart$ ();
}
if (!Clazz.isClassDefined ("org.ultrastudio.LissajousSimple.SigChart")) {
org.ultrastudio.LissajousSimple.$LissajousSimple$SigChart$ ();
}
this.fySlider = null;
this.fxSlider = null;
this.aySlider = null;
this.axSlider = null;
this.sliders = null;
this.MAX_POINTS = 1000;
this.xx = null;
this.yy = null;
this.tt = null;
this.count = 1000;
Clazz.instantialize (this, arguments);
}, org.ultrastudio, "LissajousSimple", javax.swing.JApplet, javax.swing.event.ChangeListener);
Clazz.prepareFields (c$, function () {
this.fySlider =  new javax.swing.JSlider ();
this.fxSlider =  new javax.swing.JSlider ();
this.aySlider =  new javax.swing.JSlider ();
this.axSlider =  new javax.swing.JSlider ();
this.sliders =  new javax.swing.JPanel ( new java.awt.GridLayout (4, 1));
this.xx =  Clazz.newFloatArray (1000, 0);
this.yy =  Clazz.newFloatArray (1000, 0);
this.tt =  Clazz.newFloatArray (1000, 0);
});
Clazz.makeConstructor (c$, 
function () {
Clazz.superConstructor (this, org.ultrastudio.LissajousSimple, []);
this.setLayout ( new java.awt.BorderLayout (4, 4));
this.fySlider.setValue (2);
this.fxSlider.setValue (3);
this.fySlider.setMaximum (10);
this.aySlider.setMaximum (180);
this.fxSlider.setMaximum (10);
this.axSlider.setMaximum (180);
this.fxSlider.setMinimum (1);
this.fySlider.setMinimum (1);
this.fxSlider.setMinorTickSpacing (1);
this.fySlider.setMinorTickSpacing (1);
this.aySlider.setValue (0);
this.axSlider.setValue (120);
this.axSlider.setMinorTickSpacing (10);
this.aySlider.setMinorTickSpacing (10);
this.aySlider.setMajorTickSpacing (60);
this.axSlider.setMajorTickSpacing (60);
this.fxSlider.setMajorTickSpacing (1);
this.fySlider.setMajorTickSpacing (1);
this.fxSlider.setBorder (javax.swing.BorderFactory.createTitledBorder ("First frequency, Hz"));
this.axSlider.setBorder (javax.swing.BorderFactory.createTitledBorder ("First phase, degrees"));
this.fySlider.setBorder (javax.swing.BorderFactory.createTitledBorder ("Second frequency, Hz"));
this.aySlider.setBorder (javax.swing.BorderFactory.createTitledBorder ("Second phase, degrees"));
var lissajousChart = Clazz.innerTypeInstance (org.ultrastudio.LissajousSimple.LisChart, this, null);
lissajousChart.setBorder (javax.swing.BorderFactory.createLineBorder (java.awt.Color.BLACK));
var signalChart = Clazz.innerTypeInstance (org.ultrastudio.LissajousSimple.SigChart, this, null);
signalChart.setPreferredSize ( new java.awt.Dimension (500, 100));
signalChart.setBorder (javax.swing.BorderFactory.createLineBorder (java.awt.Color.BLACK));
this.setupSlider (this.fxSlider);
this.setupSlider (this.fySlider);
this.setupSlider (this.axSlider);
this.setupSlider (this.aySlider);
this.add (signalChart, "South");
this.add (lissajousChart, "Center");
this.add (this.sliders, "East");
this.createLissajousDataSet ();
});
Clazz.defineMethod (c$, "setupSlider", 
 function (slider) {
slider.setPaintLabels (true);
slider.setPaintTicks (true);
slider.setPaintTrack (true);
slider.setSnapToTicks (true);
slider.setPreferredSize ( new java.awt.Dimension (200, 30));
this.sliders.add (slider);
slider.addChangeListener (this);
}, "javax.swing.JSlider");
Clazz.defineMethod (c$, "createLissajousDataSet", 
function () {
var fx = this.fxSlider.getValue () / 100.0;
var fy = this.fySlider.getValue () / 100.0;
var ax = Math.toRadians (this.aySlider.getValue ());
var ay = Math.toRadians (this.axSlider.getValue ());
this.xx =  Clazz.newFloatArray (1000, 0);
this.yy =  Clazz.newFloatArray (1000, 0);
var p2 = (6.283185307179586);
for (this.count = 0; this.count < this.xx.length; this.count++) {
var ff = 100.0 * 6.2831855 * (this.count) / 1000;
this.yy[this.count] = Math.sin (ax + fx * ff);
this.xx[this.count] = Math.sin (ay + fy * ff);
this.tt[this.count] = 10.0 * ff / 6.2831855;
}
if (this.count < 999) this.count++;
});
Clazz.overrideMethod (c$, "stateChanged", 
function (e) {
this.createLissajousDataSet ();
this.repaint (50);
}, "javax.swing.event.ChangeEvent");
c$.$LissajousSimple$LisChart$ = function () {
Clazz.pu$h(c$);
c$ = Clazz.decorateAsClass (function () {
Clazz.prepareCallback (this, arguments);
this.scale = 0;
Clazz.instantialize (this, arguments);
}, org.ultrastudio.LissajousSimple, "LisChart", javax.swing.JPanel);
Clazz.overrideMethod (c$, "paint", 
function (a) {
var b = a;
var c = this.getWidth ();
var d = this.getHeight ();
var e = Math.min (c, d);
this.scale = e / 2.2;
b.setStroke ( new java.awt.BasicStroke (2));
b.setColor (java.awt.Color.RED);
var f =  new java.awt.geom.Line2D.Float ();
f.x2 = this.g (this.b$["org.ultrastudio.LissajousSimple"].xx[0]);
f.y2 = this.g (this.b$["org.ultrastudio.LissajousSimple"].yy[0]);
for (var g = 1; g < this.b$["org.ultrastudio.LissajousSimple"].count; g++) {
f.x1 = f.x2;
f.y1 = f.y2;
f.x2 = this.g (this.b$["org.ultrastudio.LissajousSimple"].xx[g]);
f.y2 = this.g (this.b$["org.ultrastudio.LissajousSimple"].yy[g]);
b.draw (f);
}
}, "java.awt.Graphics");
Clazz.defineMethod (c$, "g", 
function (a) {
return (a + 1.1) * this.scale;
}, "~N");
c$ = Clazz.p0p ();
};
c$.$LissajousSimple$SigChart$ = function () {
Clazz.pu$h(c$);
c$ = Clazz.decorateAsClass (function () {
Clazz.prepareCallback (this, arguments);
this.sx = 0;
this.sy = 0;
Clazz.instantialize (this, arguments);
}, org.ultrastudio.LissajousSimple, "SigChart", javax.swing.JPanel);
Clazz.overrideMethod (c$, "paint", 
function (a) {
var b = a;
var c = this.getWidth ();
var d = this.getHeight ();
var e = (6.283185307179586);
var f = 100.0 * e * (this.b$["org.ultrastudio.LissajousSimple"].count) / 1000;
this.sx = c / f;
this.sy = d / 2.2;
b.setStroke ( new java.awt.BasicStroke (2));
b.setColor (java.awt.Color.RED);
this.paintSignal (b, this.b$["org.ultrastudio.LissajousSimple"].xx);
b.setColor (java.awt.Color.BLUE);
this.paintSignal (b, this.b$["org.ultrastudio.LissajousSimple"].yy);
}, "java.awt.Graphics");
Clazz.defineMethod (c$, "paintSignal", 
 function (a, b) {
var c =  new java.awt.geom.Line2D.Float ();
c.x2 = this.gx (this.b$["org.ultrastudio.LissajousSimple"].tt[0]);
c.y2 = this.gy (b[0]);
for (var d = 1; d < this.b$["org.ultrastudio.LissajousSimple"].count; d++) {
c.x1 = c.x2;
c.y1 = c.y2;
c.x2 = this.gx (this.b$["org.ultrastudio.LissajousSimple"].tt[d]);
c.y2 = this.gy (b[d]);
a.draw (c);
}
}, "java.awt.Graphics2D,~A");
Clazz.defineMethod (c$, "gy", 
function (a) {
return (a + 1.1) * this.sy;
}, "~N");
Clazz.defineMethod (c$, "gx", 
function (a) {
return a * this.sx;
}, "~N");
c$ = Clazz.p0p ();
};
});

File: sources/net.sf.j2s.core/test/src/org/ultrastudio/LissajousSimple.java
package org.ultrastudio;

import java.awt.BasicStroke;
import java.awt.BorderLayout;
import java.awt.Color;
import java.awt.Dimension;
import java.awt.Graphics;
import java.awt.Graphics2D;
import java.awt.GridLayout;
import java.awt.geom.Line2D;

import javax.swing.BorderFactory;
import javax.swing.JApplet;
import javax.swing.JComponent;
import javax.swing.JPanel;
import javax.swing.JSlider;
import javax.swing.event.ChangeEvent;
import javax.swing.event.ChangeListener;

//web_Ready
//web_AppletName= Lissajous
//web_Description= llustrates the Lissajous pattern (two AC signals plotted on X and Y axes)
//web_JavaVersion= http://ultrastudio.org/en/Lissajous_curves
//web_JavaSource= https://java.net/projects/ultrastudio/sources/applets/show/us_Lissajous_s
//web_AppletImage= lissajous.png
//web_Category= Physics - Electronics
//web_Date= $Date: 2017-01-01 00:08:18 -0600 (Sun, 01 Jan 2017) $
//web_Features= graphics

/**
 * Produciton allpet to draw Lissajous curves. Faster but more difficult to
 * understand.
 * 
 * @author Audrius Meskauskas
 */
@SuppressWarnings("serial")
public class LissajousSimple extends JApplet implements ChangeListener {

	final float acceleration = 100;
	
	/**
	 * The component that draws Lissajous chart.
	 */
	class LisChart extends JPanel {
		
		float scale;
		
		/**
		 * Paint the Lissajous curve.
		 */
		public void paint(Graphics gg) {
			Graphics2D g = (Graphics2D) gg;
			int w = getWidth();
			int h = getHeight();
			float min = Math.min(w, h);
			
			scale = min/2.2f;
			g.setStroke(new BasicStroke(2));
			g.setColor(Color.RED);

			Line2D.Float line = new Line2D.Float();
			line.x2 = g(xx[0]);
			line.y2 = g(yy[0]);
			for (int i = 1; i < count; i++) {
				line.x1 = line.x2;
				line.y1 = line.y2;
				line.x2 = g(xx[i]);
				line.y2 = g(yy[i]);
				g.draw(line);
			}
		}
		
		/**
		 * Convert from model to screen coordinates.
		 */
		final float g(float v) {
			return (v + 1.1f) * scale;
		}
	}
	
	/**
	 * The component that draws the signal curves chart.
	 */
	class SigChart extends JPanel {
		
		/**
		 * Horizontal scale
		 */
		float sx;
		
		/**
		 * Vertical scale.
		 */
		float sy;
		
		/**
		 * Paint the signal curves.
		 */
		public void paint(Graphics gg) {
			Graphics2D g = (Graphics2D) gg;
			float w = getWidth();
			float h = getHeight();
			
			float p2 = (float) (Math.PI * 2.0f);			
			float ww = acceleration * p2 * ((float) count) / MAX_POINTS;
			sx = w / ww;
			sy = h / 2.2f;

			g.setStroke(new BasicStroke(2));
			g.setColor(Color.RED);
			paintSignal(g, xx);
			g.setColor(Color.BLUE);
			paintSignal(g, yy);
		}

		private void paintSignal(Graphics2D g, float[] vv) {
			Line2D.Float line = new Line2D.Float();
			line.x2 = gx(tt[0]);
			line.y2 = gy(vv[0]);
			for (int i = 1; i < count; i++) {
				line.x1 = line.x2;
				line.y1 = line.y2;
				line.x2 = gx(tt[i]);
				line.y2 = gy(vv[i]);
				g.draw(line);
			}
		}
		
		/**
		 * Convert from model to screen coordinates.
		 */
		final float gy(float v) {
			return (v + 1.1f) * sy;
		}
		/**
		 * Convert from model to screen coordinates.
		 */
		final float gx(float v) {
			return v * sx;
		}
	}	

	/**
	 * The slider for F1
	 */
	JSlider fySlider = new JSlider();

	/**
	 * The slider for F2
	 */
	JSlider fxSlider = new JSlider();

	/**
	 * The slider for F1
	 */
	JSlider aySlider = new JSlider();

	/**
	 * The slider for F2
	 */
	JSlider axSlider = new JSlider();

	JPanel sliders = new JPanel(new GridLayout(4, 1));

	/**
	 * The max number of points in the charts.
	 */
	final int MAX_POINTS = 1000;
	
	float[] xx = new float[MAX_POINTS];
	float[] yy = new float[MAX_POINTS];
	float[] tt = new float[MAX_POINTS];

	/**
	 * The actual number of points that depends on where the Lissajous curve
	 * closes.
	 */
	int count = MAX_POINTS;

	/**
	 * A demonstration application showing a normal distribution.
	 * 
	 * @param title
	 *            the frame title.
	 */
	public LissajousSimple() {
		
		setLayout(new BorderLayout(4,4));

		fySlider.setValue(2);
		fxSlider.setValue(3);

		fySlider.setMaximum(10);
		aySlider.setMaximum(180);
		fxSlider.setMaximum(10);
		axSlider.setMaximum(180);
		
		fxSlider.setMinimum(1);
		fySlider.setMinimum(1);
		fxSlider.setMinorTickSpacing(1);
		fySlider.setMinorTickSpacing(1);

		aySlider.setValue(0);
		axSlider.setValue(120);
		axSlider.setMinorTickSpacing(10);
		aySlider.setMinorTickSpacing(10);		
		aySlider.setMajorTickSpacing(60);
		axSlider.setMajorTickSpacing(60);

		fxSlider.setMajorTickSpacing(1);
		fySlider.setMajorTickSpacing(1);

		fxSlider.setBorder(BorderFactory
				.createTitledBorder("First frequency, Hz"));
		axSlider.setBorder(BorderFactory
				.createTitledBorder("First phase, degrees"));
		fySlider.setBorder(BorderFactory
				.createTitledBorder("Second frequency, Hz"));
		aySlider.setBorder(BorderFactory
				.createTitledBorder("Second phase, degrees"));
		
		LisChart lissajousChart = new LisChart();
		lissajousChart.setBorder(BorderFactory.createLineBorder(Color.BLACK));
		
		SigChart signalChart = new SigChart();
		signalChart.setPreferredSize(new Dimension(500, 100));
		signalChart.setBorder(BorderFactory.createLineBorder(Color.BLACK));

		setupSlider(fxSlider);
		setupSlider(fySlider);		
		setupSlider(axSlider);		
		setupSlider(aySlider);
		add(signalChart, BorderLayout.SOUTH);
		add(lissajousChart, BorderLayout.CENTER);
		add(sliders, BorderLayout.EAST);
		createLissajousDataSet();
	}

	private void setupSlider(JSlider slider) {
		slider.setPaintLabels(true);
		slider.setPaintTicks(true);
		slider.setPaintTrack(true);
		slider.setSnapToTicks(true);
		slider.setPreferredSize(new Dimension(200,30));
		sliders.add(slider);
		slider.addChangeListener(this);
	}

	/**
	 * Creates a dataset with sample values from the normal distribution
	 * function.
	 * 
	 * @return A dataset, contains Lissajous curve and two value over time
	 *         frequencies.
	 */
	public void createLissajousDataSet() {
		float fx = fxSlider.getValue() / 100.0f;
		float fy = fySlider.getValue() / 100.0f;

		float ax = (float) Math.toRadians(aySlider.getValue());
		float ay = (float) Math.toRadians(axSlider.getValue());

		xx = new float[MAX_POINTS];
		yy = new float[MAX_POINTS];

		final float p2 = (float) (Math.PI * 2.0f);

		for (count = 0; count < xx.length; count++) {
			float ff = acceleration * p2 * ((float) count) / MAX_POINTS;
			yy[count] = (float) Math.sin(ax + fx * ff);
			xx[count] = (float) Math.sin(ay + fy * ff);
			tt[count] = 10.0f * ff / p2;
		}

		if (count < MAX_POINTS - 1)
			count++;
	}

	@Override
	/**
	 * Called whenever sliders change the position, as we registered this applet as the change listener for the sliders.
	 */
	public void stateChanged(ChangeEvent e) {
		createLissajousDataSet();
		repaint(50);
	}

}
